File: Java/datamakeexit.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.PrintWriter;

/**
 * Created by Sung Kim on 3/2/2016.
 # datamakeexit.java: ensure that all vehicles get off the xway
 # run after dataval.java and datarm2.java
 # Usage: java datamakeexit <file>  <outfile>
 # Note: NOT using redirection since System.out.println() is VERY slow

 */
public class datamakeexit {
    static final int MAX_CARID = 300000;

    public static void main(String[] args) throws Exception {
        long st = System.currentTimeMillis();
        String input_file_name = args[0];
        String output_file_name = args[1];
        int[] lasttimes = new int[MAX_CARID];
        File input_file = new File(input_file_name);
        File output_file = new File(output_file_name);
        BufferedReader reader;
        PrintWriter writer;
        String line;
        String[] tokens;
        StringBuilder out_token = new StringBuilder();

        reader = new BufferedReader(new FileReader(input_file));

        while ((line = reader.readLine()) != null) {
            tokens = line.split(",");
            int carid = Integer.parseInt(tokens[2]);
            int time = Integer.parseInt(tokens[1]);
            lasttimes[carid] = time;
        }
        reader.close();
        System.err.println("Time for first read: " + (System.currentTimeMillis() - st));

        st = System.currentTimeMillis();
        reader = new BufferedReader(new FileReader(input_file));
        writer = new PrintWriter(output_file);
        while ((line = reader.readLine()) != null) {
            tokens = line.split(",");
            int carid = Integer.parseInt(tokens[2]);
            int time = Integer.parseInt(tokens[1]);
            if (time == lasttimes[carid] && tokens[0].equals("0")) {
                tokens[3] = "10";
                tokens[5] = "4";
            }
            dataval.printTokensToFile(writer, out_token, tokens);
        }
        reader.close();
        writer.close();
        System.err.println("Time for second read: " + (System.currentTimeMillis() - st));
    }
}

File: C/datamakeexit.c
#include <stdio.h>
#include <string.h>


int MAX_CARID = 300000;
int NUM_FIELDS = 15;
int LINE_BUF_SIZE = 100;

// datamakeexit.c: ensure that all vehicles get off the xway
//  run after dataval.c and datarm2.c
// Usage: datamakeexit <file> > <outfile>

void print_tokens(int* tokens, int size) {
	int i;
	for (i = 0; i < size ; i++) { 
		if (i == size - 1) {
			printf("%d", tokens[i]);
		} else {
			printf("%d,", tokens[i]);
		}
	}
	printf("\n");
}

void tokenize(int* tokens, char* line) {
	char* token;
	int tokencount = 0;
	token = strtok(line, ",");
	while(token != NULL) {
		tokens[tokencount] = atoi(token);
		token = strtok(NULL, ",");
		tokencount++;
	}
}
			
int main(int argc, char* argv[]) {
	FILE* fp;
	int i;
	int tokens[NUM_FIELDS];
	char line[LINE_BUF_SIZE];

	int lasttimes[MAX_CARID];

	fp = fopen(argv[1], "r");
	if (fp == NULL) {
		printf("Bad file handle\n");
		return(1);
	}

	while(fgets(line, LINE_BUF_SIZE, fp) != NULL) {
		tokenize(tokens, line);	
		lasttimes[tokens[2]] = tokens[1];
	} 
	fclose(fp);
	
	fp = fopen(argv[1], "r");
	if (fp == NULL) {
		printf("Bad file handle\n");
		return(1);
	}

	while(fgets(line, LINE_BUF_SIZE, fp) != NULL) {
		tokenize(tokens, line);	
		if (tokens[1] == lasttimes[tokens[2]] && tokens[0] == 0) { // only last appearing type 0 queries need adjustment 
			tokens[3] = 10;
			tokens[5] = 4;
			print_tokens(tokens, NUM_FIELDS);	
		} else {
			print_tokens(tokens, NUM_FIELDS);	
		}
	} 
	

	return(0);
}

